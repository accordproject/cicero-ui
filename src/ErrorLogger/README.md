## Cicero-UI - ErrorLogger

### Usage

```
npm install @accordproject/cicero-ui
```

```js
import { ErrorLogger } from '@accordproject/cicero-ui';

const navigateToClauseError = // Insert your navigation function here.

const errorsPropsObject = {
    ERRORS_HEADER_BACKGROUND (string),
    ERRORS_HEADER_BACKGROUND_HOVER (string),
    ERRORS_HEADER_EXPAND_ARROW (string),
    ERRORS_HEADER_BORDER_TOP (string),
    ERRORS_HEADER_SHADOW (string),
    ERRORS_DISPLAY_BACKGROUND (string),
    ERRORS_DISPLAY_SHADOW (string),
    ERROR_BORDER_BOTTOM (string),
    ERROR_EXPAND_ARROW (string),
    ERROR_FILE (string),
    ERROR_FILE_HOVER (string),
    ERROR_TYPE (string),
    ERROR_FULL_MESSAGE (string),
    ERROR_SHORT_MESSAGE (string),
};

const ErrorContainer = props => (
    <ErrorLogger errors={props.errors} errorsProps={errorsPropsObject} errorNav={navigateToClauseError}/>
);
```

### Props

#### Expected to be Provided

- `errors` : An `array` which contains error objects.

#### Require Provision

- `errorsProps` : An optional `object`, see below.
- `errorNav` : An optional `function` which will navigate to the corresponding error.

### Specifications

This component is built to have the following dimensions:

```js
width: '100%';
```

You can style the error component, as well as the individual errors. An object may be passed down this component with the following possible CSS inputs as strings:
- `ERRORS_HEADER_BACKGROUND`
- `ERRORS_HEADER_BACKGROUND_HOVER`
- `ERRORS_HEADER_EXPAND_ARROW`
- `ERRORS_HEADER_BORDER_TOP`
- `ERRORS_HEADER_SHADOW`
- `ERRORS_DISPLAY_BACKGROUND`
- `ERRORS_DISPLAY_SHADOW`
- `ERROR_BORDER_BOTTOM`
- `ERROR_EXPAND_ARROW`
- `ERROR_FILE`
- `ERROR_FILE_HOVER`
- `ERROR_TYPE`
- `ERROR_FULL_MESSAGE`
- `ERROR_SHORT_MESSAGE`