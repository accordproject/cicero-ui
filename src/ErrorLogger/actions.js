export const gtZero = input => input > 0;

export const typeSwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return 'Model';
    case ('parseError' in input):
      return 'Grammar';
    default:
      return 'Unknown';
  }
};

export const keySwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return input.clauseTemplateId;
    case ('parseError' in input):
      return input.clauseId;
    default:
      return null;
  }
};

export const overalltypeSwitchCase = (input) => {
  switch (true) {
    case ('modelError' in input):
      return input.modelError;
    case ('parseError' in input):
      return input.parseError;
    default:
      return null;
  }
};

export const errorsExist = errors => gtZero(errors.length);

export const errorArrayLength = errors => (errorsExist(errors) ? errors.length : 'No');

export const isMultpleErrors = errors => ((errors.length > 1) ? 'Errors' : 'Error');

export const truncateMessage = string => ((string.length > 200)
  ? `${string.substring(0, 200)}...` : string);
