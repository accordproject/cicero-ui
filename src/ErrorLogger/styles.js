/* Styling */
import styled from 'styled-components';
import { Icon } from 'semantic-ui-react';

/* Components */
export const ErrorDisplay = styled.div`
    bottom: 25px;
    width: 100%;
    position: fixed;
    max-height: 300px;
    background-color: ${props => props.displayBackground || '#1E2D53'};
    box-shadow: ${props => props.displayShadow || '0 -2px 20px 0 rgba(20,31,60,0.65)'};
    z-index: ${props => props.zIndexInput || 'auto'};
`;

export const ErrorsHeader = styled.div`
    z-index: ${props => props.zIndexInput || 'auto'};
    width: 100%;
    position: fixed;
    bottom: ${props => (props.errors ? '0' : 'auto')};
    transition: 1s;
    height: ${props => (props.errors ? '25px' : '0')};
    padding: 0.1em 0.1em 0.1em 1em;
    background-color: ${props => props.headerBackground || '#1E2D53'};
    box-shadow: ${props => props.headerShadow || '0 -2px 20px 0 rgba(20,31,60,0.65)'};
    border-top: ${props => props.headerTop || ' 1px solid #50637F'};

    color: ${props => (props.errors ? '#FF4242' : '#19C6C7')};
    font-family: "IBM Plex Sans";
    font-size: 1em;
    font-weight: bold;
    letter-spacing: -0.5px;
    line-height: 20px;

    &:hover {
        background-color: ${props => props.headerBackgroundHover || '#364C77'};
        cursor: pointer;
    }
`;

export const ErrorSymbol = styled(Icon)`
    vertical-align: middle;
`;

export const ErrorBarArrow = styled.div`
    float: right;
    margin: 5px 15px;

    border-top: ${props => (props.errorDisplay
    ? (`7px solid ${props.headerBarArrow || '#7B9AD1'}`) : '0')};

    border-right: 4px solid transparent;

    border-left: 4px solid transparent;

    border-bottom: ${props => (props.errorDisplay
    ? '0' : (`7px solid ${props.headerBarArrow || '#7B9AD1'}`))};
`;

export const ErrorComponent = styled.div`
    width: 100%;
    color: #F0F0F0;
    border-bottom: 1px solid ${props => props.borderBottom || '#50637F'};
    padding: 10px 16px;

    display: grid;
    grid-row-gap: 20px; 
    grid-template-areas: "errorArrow errorFile errorType errorMessage"
                        "errorFull errorFull errorFull errorFull";
    grid-template-columns: 0.25fr 1fr 1fr 8fr;
    grid-template-rows: min-content auto;
`;

export const ErrorFile = styled.a`
    text-decoration: underline;

    color: ${props => props.errorFile || '#FFFFFF'};
    font-family: "IBM Plex Sans";
    font-size: 0.81em;
    line-height: 13px;

    grid-area: errorFile;
    align-self: center;
    &:hover {
        cursor: pointer;
        color: ${props => props.errorFileHover || '#0066CC'};
    }
`;

export const ErrorType = styled.div`
    grid-area: errorType;
    color: ${props => props.errorType || '#B9BCC4'};
    font-family: "IBM Plex Sans";
    font-size: 0.81em;
    line-height: 13px;
    align-self: center;
    padding: 5px;
`;

export const ErrorShortMessage = styled.div`
    grid-area: errorMessage;
    color: ${props => props.shortMessage || '#B9BCC4'};
    font-family: "IBM Plex Sans";
    font-size: 0.81em;
    line-height: 13px;
    align-self: center;
    padding: 5px;
`;

export const ErrorFullMessage = styled.div`
    grid-area: errorFull;
    color: ${props => props.fullMessage || '#FFFFFF'};
    font-family: "IBM Plex Sans";
    font-size: 0.81em;
    line-height: 13px;
`;

export const ArrowDiv = styled.div`
    grid-area: errorArrow;
    place-self: center;
    width: 0; 
    height: 0; 
    margin: 5px;

    border-top: ${props => (props.expanded
    ? (`10px solid ${props.errorArrow || '#50637F'}`) : '4px solid transparent')};

    border-right: ${props => (props.expanded ? '4px solid transparent' : '0')};

    border-bottom: ${props => (props.expanded ? '0' : '4px solid transparent')};

    border-left: ${props => (props.expanded
    ? '4px solid transparent' : (`10px solid ${props.errorArrow || '#50637F'}`))};

    &:hover {
        cursor: pointer;
    }
`;
