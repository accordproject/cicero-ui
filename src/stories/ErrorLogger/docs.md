# Error Logger

#### Notes for Error Logger

---

Supports code snippets too:

```jsx
<div>
  Foo
</div>
```