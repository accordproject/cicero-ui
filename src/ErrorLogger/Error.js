/* React */
import React, { useState } from 'react';
import PropTypes from 'prop-types';

/* Actions */
import * as ACT from './actions';

/* Styling */
import * as SC from './styles';

const ErrorComponent = (props) => {
  const { error, errorProps, errorNav } = props;
  const [specErrorVisible, setspecErrorVisible] = useState(false);

  const handleClickSpecError = () => {
    setspecErrorVisible(!specErrorVisible);
  };

  const componentProps = {
    borderBottom: errorProps.ERROR_BORDER_BOTTOM,
  };

  const fileProps = {
    errorFile: errorProps.ERROR_FILE,
    errorFileHover: errorProps.ERROR_FILE_HOVER,
  };

  const typeProps = {
    onClick: handleClickSpecError,
    errorType: errorProps.ERROR_TYPE,
  };

  const shortMessageProps = {
    onClick: handleClickSpecError,
    shortMessage: errorProps.ERROR_SHORT_MESSAGE,
  };

  const fullMessageProps = {
    fullMessage: errorProps.ERROR_FULL_MESSAGE,
  };

  const errorArrowProps = {
    expanded: specErrorVisible,
    onClick: handleClickSpecError,
    errorArrow: errorProps.ERROR_EXPAND_ARROW,
  };

  return (
    <SC.ErrorComponent {...componentProps}>

      <SC.ArrowDiv {...errorArrowProps} />
      <SC.ErrorFile {...fileProps} onClick={() => errorNav(error)} >
        {ACT.typeSwitchCase(error || {})}
      </SC.ErrorFile>

      <SC.ErrorType {...typeProps} >
        {ACT.overalltypeSwitchCase(error).name || 'Unknown Error'}:
      </SC.ErrorType>

      <SC.ErrorShortMessage {...shortMessageProps} >
        {ACT.truncateMessage(ACT.overalltypeSwitchCase(error).shortMessage || 'Unknown Error')}
      </SC.ErrorShortMessage>

      {specErrorVisible
        && <SC.ErrorFullMessage {...fullMessageProps} >
            {ACT.overalltypeSwitchCase(error).message || 'Unknown Error'}
          </SC.ErrorFullMessage>}
    </SC.ErrorComponent>
  );
};

ErrorComponent.propTypes = {
  error: PropTypes.object.isRequired,
  errorNav: PropTypes.func,
  errorProps: PropTypes.shape({
    ERRORS_HEADER_BACKGROUND: PropTypes.string,
    ERRORS_HEADER_BACKGROUND_HOVER: PropTypes.string,
    ERRORS_HEADER_EXPAND_ARROW: PropTypes.string,
    ERRORS_HEADER_BORDER_TOP: PropTypes.string,
    ERRORS_HEADER_SHADOW: PropTypes.string,
    ERRORS_DISPLAY_BACKGROUND: PropTypes.string,
    ERRORS_DISPLAY_SHADOW: PropTypes.string,
    ERROR_BORDER_BOTTOM: PropTypes.string,
    ERROR_EXPAND_ARROW: PropTypes.string,
    ERROR_FILE: PropTypes.string,
    ERROR_FILE_HOVER: PropTypes.string,
    ERROR_TYPE: PropTypes.string,
    ERROR_FULL_MESSAGE: PropTypes.string,
    ERROR_SHORT_MESSAGE: PropTypes.string,
  }),
};

export default ErrorComponent;
