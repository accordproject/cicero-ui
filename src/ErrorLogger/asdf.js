/* React */
import React, { useState } from 'react';
import PropTypes from 'prop-types';

/* Actions */
import * as A from './actions';

/* Styling */
import * as S from './styles';

const ErrorComponent = (props) => {
  const { error, errorProps, errorNav } = props;

  const [specErrorVisible, setspecErrorVisible] = useState(false);

  const handleClickSpecError = () => {
    setspecErrorVisible(!specErrorVisible);
  };

  const componentProps = {
    borderBottom: errorProps.ERROR_BORDER_BOTTOM,
  };

  const fileProps = {
    errorFile: errorProps.ERROR_FILE,
    errorFileHover: errorProps.ERROR_FILE_HOVER,
  };

  const typeProps = {
    onClick: handleClickSpecError,
    errorType: errorProps.ERROR_TYPE,
  };

  const shortMessageProps = {
    onClick: handleClickSpecError,
    shortMessage: errorProps.ERROR_SHORT_MESSAGE,
  };

  const fullMessageProps = {
    fullMessage: errorProps.ERROR_FULL_MESSAGE,
  };

  const errorArrowProps = {
    expanded: specErrorVisible,
    onClick: handleClickSpecError,
    errorArrow: errorProps.ERROR_EXPAND_ARROW,
  };

  return (
    <S.ErrorComponent {...componentProps}>

      <S.ArrowDiv {...errorArrowProps} />
      <S.ErrorFile {...fileProps} onClick={() => errorNav(error)} >
        {A.typeSwitchCase(error)}
      </S.ErrorFile>

      <S.ErrorType {...typeProps} >
        {A.overalltypeSwitchCase(error).name}:
      </S.ErrorType>

      <S.ErrorShortMessage {...shortMessageProps} >
        {A.truncateMessage(A.overalltypeSwitchCase(error).shortMessage)}
      </S.ErrorShortMessage>

      {specErrorVisible
        && <S.ErrorFullMessage {...fullMessageProps} >
            {A.overalltypeSwitchCase(error).message}
          </S.ErrorFullMessage>}
    </S.ErrorComponent>
  );
};

ErrorComponent.propTypes = {
  error: PropTypes.object.isRequired,
  errorNav: PropTypes.func,
  errorProps: PropTypes.shape({
    ERRORS_HEADER_BACKGROUND: PropTypes.string,
    ERRORS_HEADER_BACKGROUND_HOVER: PropTypes.string,
    ERRORS_HEADER_EXPAND_ARROW: PropTypes.string,
    ERRORS_HEADER_BORDER_TOP: PropTypes.string,
    ERRORS_HEADER_SHADOW: PropTypes.string,
    ERRORS_DISPLAY_BACKGROUND: PropTypes.string,
    ERRORS_DISPLAY_SHADOW: PropTypes.string,
    ERROR_BORDER_BOTTOM: PropTypes.string,
    ERROR_EXPAND_ARROW: PropTypes.string,
    ERROR_FILE: PropTypes.string,
    ERROR_FILE_HOVER: PropTypes.string,
    ERROR_TYPE: PropTypes.string,
    ERROR_FULL_MESSAGE: PropTypes.string,
    ERROR_SHORT_MESSAGE: PropTypes.string,
  }),
};

export default ErrorComponent;
