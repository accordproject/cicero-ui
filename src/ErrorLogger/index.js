/* React */
import React, { useState } from 'react';
import PropTypes from 'prop-types';

/* Styling */
import * as ACT from './actions';

/* Actions */
import * as SC from './styles';

/* Component */
import ErrorComponent from './Error';

const ErrorLogger = (props) => {
  const { errors, errorNav } = props;
  const errorsProps = props.errorsProps || Object.create(null);

  const [errorsVisible, setErrorsVisible] = useState(false);

  const handleClickErrorsBar = () => {
    if (ACT.gtZero(errors.length)) { setErrorsVisible(!errorsVisible); }
  };

  const headerProps = {
    id: 'ErrorComponentHeader',
    errors: ACT.errorsExist(errors),
    onClick: handleClickErrorsBar,
    headerBackground: errorsProps.ERRORS_HEADER_BACKGROUND,
    headerBackgroundHover: errorsProps.ERRORS_HEADER_BACKGROUND_HOVER,
    headerShadow: errorsProps.ERRORS_HEADER_SHADOW,
    headerTop: errorsProps.ERRORS_HEADER_BORDER_TOP,
    zIndexInput: errorsProps.ERRORS_DISPLAY_Z_INDEX,
  };

  const displayProps = {
    id: 'ErrorComponentDisplay',
    errorDisplay: errorsVisible,
    displayBackground: errorsProps.ERRORS_DISPLAY_BACKGROUND,
    displayShadow: errorsProps.ERRORS_DISPLAY_SHADOW,
    zIndexInput: errorsProps.ERRORS_DISPLAY_Z_INDEX,
  };

  const barArrowProps = {
    errorDisplay: errorsVisible,
    headerBarArrow: errorsProps.ERRORS_HEADER_EXPAND_ARROW,
  };

  const symbolProps = {
    name: 'exclamation triangle',
    size: 'small'
  };

  const errorComponentGenerator = errors => errors
    .map(soloError => <ErrorComponent
      errorProps={errorsProps}
      error={soloError}
      errorNav={errorNav}
      key={ACT.keySwitchCase(soloError)} />);

  return (
    <div>
      {errorsVisible
        && <SC.ErrorDisplay {...displayProps} >
            {errorComponentGenerator(errors)}
          </SC.ErrorDisplay>}

      <SC.ErrorsHeader {...headerProps} >
        {ACT.gtZero(errors.length)
          && <SC.ErrorSymbol {...symbolProps} />}
        {ACT.errorArrayLength(errors)} {ACT.isMultipleErrors(errors)}
        <SC.ErrorBarArrow {...barArrowProps} />
      </SC.ErrorsHeader>
    </div>
  );
};

ErrorLogger.propTypes = {
  errors: PropTypes.array.isRequired,
  errorNav: PropTypes.func,
  errorsProps: PropTypes.shape({
    ERRORS_HEADER_BACKGROUND: PropTypes.string,
    ERRORS_HEADER_BACKGROUND_HOVER: PropTypes.string,
    ERRORS_HEADER_EXPAND_ARROW: PropTypes.string,
    ERRORS_HEADER_BORDER_TOP: PropTypes.string,
    ERRORS_HEADER_SHADOW: PropTypes.string,
    ERRORS_DISPLAY_BACKGROUND: PropTypes.string,
    ERRORS_DISPLAY_SHADOW: PropTypes.string,
    ERRORS_DISPLAY_Z_INDEX: PropTypes.string,
    ERROR_BORDER_BOTTOM: PropTypes.string,
    ERROR_EXPAND_ARROW: PropTypes.string,
    ERROR_FILE: PropTypes.string,
    ERROR_FILE_HOVER: PropTypes.string,
    ERROR_TYPE: PropTypes.string,
    ERROR_FULL_MESSAGE: PropTypes.string,
    ERROR_SHORT_MESSAGE: PropTypes.string,
  }),
};

export default ErrorLogger;
