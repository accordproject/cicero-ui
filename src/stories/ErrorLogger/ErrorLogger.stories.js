import React from "react";
import { Button } from "@storybook/react/demo";
import docs from "./docs.md";
import { withA11y } from "@storybook/addon-a11y";

export default { title: "'Components/Error Logger" };

export const errorLogger = () => <Button>Cicero-UI</Button>; // Please add errorLogger Component definition here

errorLogger.story = {
  component: errorLogger,
  decorators: [withA11y],

  parameters: {
    notes: docs
  }
};