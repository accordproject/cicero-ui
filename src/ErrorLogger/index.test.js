import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ErrorLogger from './index';

const parseError1 = {
  clauseId: 'b28f32b2-57d9-41e1-b871-03c9faa75b6e',
  parseError: {
    fileLocation: {
      end: {
        endColumn: 2,
        line: 1
      },
      start: {
        column: 1
      },
    },
    component: 'cicero-core',
    fileName: undefined,
    name: 'ParseException',
    shortMessage: '"invalid syntax at line 1 col 1:↵↵  a  Acceptance of Delivery. "Party A" will be deemed to have completed its delivery obligations if in "Party B"\'s opinion, the "Widgets" satisfies the Acceptance Criteria, and "Party B" notifies "Party A" in writing that it is accepting the "Widgets".↵  ^↵Unexpected "a"↵"',
    message: '"invalid syntax at line 1 col 1:↵↵  a  Acceptance of Delivery. "Party A" will be deemed to have completed its delivery obligations if in "Party B"\'s opinion, the "Widgets" satisfies the Acceptance Criteria, and "Party B" notifies "Party A" in writing that it is accepting the "Widgets".↵  ^↵Unexpected "a"↵"'
  }
};

describe('<ErrorLogger />', () => {
  describe('on initialization', () => {
    it('renders page correctly with errors', () => {
      const component = shallow(<ErrorLogger errors={[parseError1]} />);
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });

    it('renders page correctly without errors', () => {
      const component = shallow(<ErrorLogger errors={[]} />);
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
  });

  describe('renders conditional to errors', () => {
    it('with errors existing', () => {
      const component = shallow(<ErrorLogger errors={[parseError1]} />);
      expect(component.find('#ErrorComponentHeader').exists()).toBeTruthy();

      component.find('#ErrorComponentHeader').simulate('click');

      expect(component.find('#ErrorComponentDisplay').exists()).toBeTruthy();
    });

    it('without errors existing', () => {
      const component = shallow(<ErrorLogger errors={[]} />);
      expect(component.find('#ErrorComponentHeader').exists()).toBeTruthy();
      expect(component.find('#ErrorComponentDisplay').exists()).toBe(false);

      component.find('#ErrorComponentHeader').simulate('click');

      expect(component.find('#ErrorComponentDisplay').exists()).toBe(false);
    });
  });
});
