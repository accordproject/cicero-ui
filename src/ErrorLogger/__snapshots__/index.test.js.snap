// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorLogger /> on initialization renders page correctly with errors 1`] = `
<div>
  <styled.div
    errors={true}
    id="ErrorComponentHeader"
    onClick={[Function]}
  >
    <Styled(Icon)
      name="exclamation triangle"
      size="small"
    />
    1
     
    Error
    <styled.div
      errorDisplay={false}
    />
  </styled.div>
</div>
`;

exports[`<ErrorLogger /> on initialization renders page correctly without errors 1`] = `
<div>
  <styled.div
    errors={false}
    id="ErrorComponentHeader"
    onClick={[Function]}
  >
    No
     
    Error
    <styled.div
      errorDisplay={false}
    />
  </styled.div>
</div>
`;
